cpp/src/io/json/json_reader.cu

```diff
@@ -267,31 +268,33 @@ void JsonReader::setRecordStarts() {
 
 void JsonReader::uploadDataToDevice() {
   size_t start_offset = 0;
-  size_t bytes_to_upload = uncomp_size_;
+  size_t end_offset = uncomp_size_;
 
   // Trim lines that are outside range
-  if (args_->byte_range_size != 0) {
+  if (args_->byte_range_size != 0 || args_->byte_range_offset != 0) {
     std::vector<uint64_t> h_rec_starts(rec_starts_.size());
     CUDA_TRY(
         cudaMemcpy(h_rec_starts.data(), rec_starts_.data(), sizeof(uint64_t) * h_rec_starts.size(), cudaMemcpyDefault));
 
-    auto it = h_rec_starts.end() - 1;
-    while (it >= h_rec_starts.begin() && *it > args_->byte_range_size) {
-      --it;
+    if (args_->byte_range_size != 0) {
```

Comment:
Just by inspecting `uploadDataToDevice()` it's not clear to me how you're handling non-zero `byte_range_offset`.

1. Is the first entry in `rec_starts_` already adjusted based on `byte_range_offset`?
2. Since `byte_range_offset` may not align with the start of a record, don't you have to adjust `byte_range_size` accordingly?  Let's say `byte_range_offset = 100` and `byte_range_size = 200`.  You find a record start at `offset 80` so now your byte range size would be `200 + (100-80) = 220`.

Reply:
1. rec_starts_ elements are adjusted based on the byte range offset, because h_uncomp_data starts from the given offset. and record starts are computed with respect to this buffer.
2. Since we only copy data starting from the byte range offset, if the offset is not aligned with a row start, the row that starts before the offset will not be included in the output. So the first rec_starts_ element with always be after the start of the byte range (unless the byte range offset is zero).